### PR TITLE
[Dynamo] Trace enter/exit of TorchFunctionModes

### PR DESCRIPTION
**Pull request by mlazos**
This PR implements tracing of with contexts with TorchFunction modes which have the default enter/exit behavior (ie pushing/popping the mode)

Typically the bytecode for a context manager looks like this during a graph break:
1. graph call
2. enter context
3. unsupported code
4. exit context
5. resume call

resume fn structure:
1. enter context
2. jump
...
3. exit context

The issue with torch function modes is that side effects will replay any mutations to the torch function stack performed during tracing. So, we do not need to enter and exit around the unsupported code in the original function (doing so would result in a duplicate torch function mode entry during execution of the unsupported code), and we don't need to enter again in the resume function (the mode that was pushed from the side effects bytecode would still be on the stack). 

So for torch function modes the structure of our output code is this:

1. graph call
2. mutate tf mode stack to replay mutations
4. unsupported code
5. on exception restore stack
6. resume function

Then our resume fn looks like this:

1. no-op enter torch function mode
2. jump
3.  exit tf mode

To implement the no-op enter of the torch function mode I added torch function mode in polyfill which no-op enters, but normally exits. This is needed because we still want to trace the with context in the resume function, and exit properly (the exit instructions will still be in the function, so we need to generate instructions to set up the context).

Separately from the bytecode, dynamo also tracks contexts on the block stack, which is how the SETUP_* instructions are implemented. Naturally at a graph break, we exit these block stacks to properly reset the contexts entirely, so that we can re-enter around the unsupported code soundly. However once again, in the torch function mode case, in the event of a graph we do not want to perform any exit side effects because we want to preserve the state of the mode stack as is so that we will properly update the stack with bytecode mentioned in the first section. If we exited here, dynamo would pop the mode off of the symbolic stack, and not update the true python torch function mode stack with the suffix bytecode. All in all, for torch function modes we enter exactly once, update the global torch function mode stack with side effects bytecode, re-read this stack when compiling the resume function, and exit exactly once in the resume function. This matches the semantics of eager exactly.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #135503
* #135502
* __->__ #135422
* #135444
* #135443
* #133137
* #134732

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @rec

### CODE REVIEW

File: torch/_dynamo/variables/torch_function.py [hunk 1]
```diff
@@ -189,9 +199,26 @@ def get_mode_index(cls, ind):
         return ind + cls.offset
 
 
-class TorchFunctionModeVariable(ContextWrappingVariable):
+class TorchFunctionModeVariable(GenericContextWrappingVariable):
```
Comment (williamwen42) [inline torch/_dynamo/variables/torch_function.py:202]: Out of curiosity - what cases require us to use `GenericContextWrappingVariable` over `ContextWrappingVariable`?

**Review by williamwen42 — COMMENTED** [review]
LGTM, except we need to think about how we will handle unsupported ops that throw - the current implementation would be missing the `except` block in the generated bytecode, leading to incorrect behavior if the unsupported op throws.

i.e. the generated bytecode should be like
1. graph call
2. mutate tf mode stack to replay mutations
3. unsupported code
4. if exception occurred, restore tf stack, rethrow
5. resume function

File: torch/_dynamo/variables/torch_function.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/variables/torch_function.py:202]: We need the UDO behavior, sometimes there are custom methods on the mode that get called. I feel like this is happening in other places in dynamo too (ie subclassing dict) we might need to start thinking about a better way to have UDO like behavior on other types, but went with this.

File: torch/_dynamo/resume_execution.py [hunk 2]
```diff
@@ -48,6 +48,107 @@ class ReenterWith:
     stack_index: int
     target_values: Optional[Tuple[Any, ...]] = None
 
+    def try_except_torch_function_mode(self, code_options, cleanup: List[Instruction]):
+        """
+        Codegen based off of:
+        try:
+            (rest)
+        finally:
```
Comment (williamwen42) [inline torch/_dynamo/resume_execution.py:56]: It looks like this bytecode is actually doing:
```python
try:
    (rest)
except:
    torch._dynamo.utils.set_torch_function_mode_stack(___prev_torch_function_mode_stack)
```

Update comment to reflect this?

File: torch/_dynamo/side_effects.py [hunk 3]
```diff
@@ -593,11 +593,22 @@ def codegen_update_mutated(self, cg: PyCodegen):
...
                 var, variables.torch_function.TorchFunctionModeStackVariable
             ):
+                # Needed in the finally block for stack restoration
+                cg.add_push_null(
+                    lambda: cg.load_import_from(
+                        utils.__name__, "get_torch_function_mode_stack"
+                    )
+                )
+                cg.call_function(0, False)
+                name = variables.torch_function.get_prev_stack_var_name()
```
Comment (williamwen42) [inline torch/_dynamo/side_effects.py:603]: Should this name be different for each `TorchFunctionModeVariable`? If there are multiple `TorchFunctionModeVariable` contexts on the stack, we would be referring to the same object.

Comment (mlazos) [reply torch/_dynamo/side_effects.py:603]: No, there's only one TorchFunctionModeStackVariable (it's a singleton). This is used to flag in side effects if there is any mutation of the stack. If there is, we generate the bytecode to update it accordingly. For the exception handling, we'll only ever restore the previous version of stack once, so it is only stashed in the local once.

**Review by williamwen42 — APPROVED** [review]

**Pull request closed without merging**